### DISCUSSION TITLE
ProfilerDiagnosticObserver memory grows indefinitely until System.OutOfMemoryException occurs

### DISCUSSION THREAD

**Discussion by LEIRONGHUA** [ServiceStack/Discuss#127] [Q&A] [answered]
Recently, I upgraded from version 8.2.2 to 8.3.0, and the memory of the IIS Worker Process process increased indefinitely until a System.OutOfMemoryException error occurred.

In ProfilingFeature, Default Capacity is 10000 and MaxBodyLength is 1MB. When OutOfMemoryException occurs, the memory usage of IIS Worker Process is 10.0 GB (including virtual memory). Is there any memory leak problem?

**Environment:**

ServiceStack Version：8.3.0
AspNetCore Version：8.0
OS Version：Windows Server 2022 Datacenter 21H2 (OS Build Number: 20348.2227)
OS Memory Capacity：8GB
IIS Version：10.0.20348.1

**Example code/Steps to reproduce:**

```
public class ConfigureProfiling : IHostingStartup
    {
        public void Configure(IWebHostBuilder builder)
        {
            builder.ConfigureAppHost(host =>
            {
                host.Plugins.Add(new RequestLogsFeature
                {
                    EnableResponseTracking = true, 
                    EnableRequestBodyTracking = true, 
                    RequestLogger = new RedisRequestLogger(host.Container.Resolve<IRedisClientsManager>(), 1000)
                });

                
                host.Plugins.Add(new ProfilingFeature
                {
                    IncludeStackTrace = true,
                    DiagnosticEntryFilter = (entry, evt) =>
                    {
                        if (evt is RequestDiagnosticEvent requestEvent)
                        {
                            var req = requestEvent.Request;
                            entry.Meta = new()
                            {
                                ["RemoteIp"] = req.RemoteIp,
                                ["Referrer"] = req.UrlReferrer?.ToString()
                            };
                        }
                    },
                });
            });
        }
    }
```

**Output:**

![image](https://github.com/user-attachments/assets/be706caf-5823-41da-a2fd-5892bd025c34)

BTW, the memory usage of ServiceStack.Text.RecyclableMemoryStreamManager also seems to be high after I disable ProfilingFeature.
![image](https://github.com/user-attachments/assets/3ea8c684-c39d-426c-94e9-8e9305aba149)

**Comment by mythz** [comment, marked as answer]
Neither Profiling or [RecyclableMemoryStream.cs](https://github.com/ServiceStack/ServiceStack/blob/main/ServiceStack.Text/src/ServiceStack.Text/RecyclableMemoryStream.cs) has been touched for a number of releases. But it should only be used during development as it taps into and enables into a .NET's diagnostics event system and retains a rolling list of 10,000 events (overridable in ProfilingFeature.Capacity) that's used in the Admin Profiling UI. 

As such we only recommend using it during development, which you can enable with:

```csharp
host.Plugins.AddIfDebug(new ProfilingFeature {
    //...
});
```

**Reply by LEIRONGHUA** [reply]
Thank you for your recommendation, I have already done this.
the memory usage of ServiceStack.Text.RecyclableMemoryStreamManager also seems to be high after I disable ProfilingFeature.
Is the result of this picture normal?
![image](https://github.com/user-attachments/assets/92be138f-bb3e-4400-8dfe-e616336ca25a)

**Comment by mythz** [comment]
It manages a pool on buffers so it would depend on how and how much it's being used.

You can customize the pool configuration with:

```csharp
MemoryStreamFactory.RecyclableInstance = new(...);
```

Or disable its usage with:

```csharp
MemoryStreamFactory.UseRecyclableMemoryStream = false;
```
